### PR TITLE
Fix issues 17741 and 17742

### PR DESCRIPTION
**Pull request by Biotronic**
[17741](https://issues.dlang.org/show_bug.cgi?id=17741): std.range.transposed does not work with Map
[17742](https://issues.dlang.org/show_bug.cgi?id=17742): std.range.transposed does not have opIndex

### CODE REVIEW

File: std/range/package.d [hunk 1]
```diff
@@ -6590,11 +6595,53 @@ if (isForwardRange!RangeOfRanges &&
...
+        {
+            return _input.map!(a => a[start..end]).array.transposed;
+        }
+    }
 
 private:
     RangeOfRanges _input;
 }
 
+struct TransposedIndex(RangeOfRanges)
```
Comment (Biotronic) [inline std/range/package.d]: RFC: Are there better ways of doing this?

File: std/range/package.d [hunk 2]
```diff
@@ -6590,11 +6595,53 @@ if (isForwardRange!RangeOfRanges &&
     }
 
     auto opSlice() { return this; }
+    
+    static if (hasSlicing!(ElementType!RangeOfRanges))
```
Comment (Biotronic) [inline std/range/package.d]: RFC: Slicing is suddenly O(N) for a T[M][N]. Is this acceptable? Also, now that I look at the bugs again, it's not really part of the bug report. Should I remove it, file an enhancement request and make a separate PR?

Comment (JackStouffer) [inline std/range/package.d]: Range primitives must be O(1) with respect to the number of elements.

Comment (Biotronic) [inline std/range/package.d]: Right. No slicing, then. Could be done by saving offsets in the struct and having the cost per element lookup, but that complicates the code for very little benefit.

File: std/range/package.d [hunk 3]
```diff
@@ -6562,17 +6561,23 @@ if (isForwardRange!RangeOfRanges &&
...
 
-    // ElementType opIndex(size_t n)
-    // {
-    //     return _input[n].front;
-    // }
+    static if (isRandomAccessRange!RangeOfRanges && isRandomAccessRange!(ElementType!RangeOfRanges))
+    {
+        auto opIndex(size_t n)
+        {
+            return TransposedIndex!RangeOfRanges(_input, n);
```
Comment (JackStouffer) [inline std/range/package.d]: Maybe this is a stupid question, but why not just return the range at the `n`th position? That way you can still have `opIndex` even if the `ElementType!RangeOfRanges` isn't random access.

Comment (Biotronic) [inline std/range/package.d]: Since it's transposed, opIndex should return a column instead of a row. That is, one element from each range.

File: std/range/package.d (hunk 1 above)
Comment (quickfur) [inline std/range/package.d]: Since we're requiring random access on subranges in order for `opIndex` to be possible, what about doing something similar to what `.front` does? I.e., something like:

````
// Warning: untested code
return _input.save
             .filter!(a => !a.empty)
             .map!(a => a[i]);
````

**Comment by quickfur** [comment]
Also, I'm a bit wary of `opIndex` returning a different type from `.front`, but I suppose in this case it's unavoidable.

File: std/range/package.d (hunk 1 above)
Comment (Biotronic) [inline std/range/package.d]: That works with the supplied tests, but fails for `[[1,2,3],[4,5,6]].transposed[0][0] = 0;`. Admittedly, that same test fails for `[[1,2,3],[4,5,6]].transposed.front.front = 0;`, so maybe that's better?

**Comment by Biotronic** [comment]
An effect of `opIndex` and `front` returning different types is that `[[1,2,3],[4,5,6]].transposed.front.front = 0;` fails with the current implementation, which was surprising to me. My goal with the introduction of the TransposedIndex type was to make indexed assignment work.

IOW, they should return the same type - either both should fail to assign, or neither.

File: std/range/package.d [hunk 4]
```diff
@@ -6562,17 +6556,23 @@ if (isForwardRange!RangeOfRanges &&
             if (!e.empty)
             {
                 e.popFront();
-                r.front = e;
+                static if (hasAssignableElements!(ElementType!RangeOfRanges))
+                {
+                    r.front = e;
```
Comment (Biotronic) [inline std/range/package.d]: Clarification requested: What is this supposed to be doing? It's assigning `r.front`, then immediately calls `r.popFront();`. IOW, it's only heating the CPU. 

Note that I'm not asking why I wrote the code I did, but why this line exists in the previous incarnation of this code.

**Comment by quickfur** [comment]
If it's possible to make them the same type (without undue compromises), IMO that would be better.

File: std/range/package.d (hunk 1 above)
Comment (wilzbach) [inline std/range/package.d]: In any case it should be `private`  for now

**Comment by andralex** [comment]
Status of this? The issues have been since addressed.

**Comment by Biotronic** [comment]
As you say, it's been fixed. Closing.
